Alerting with Loki tutorial (#72704)
grafana/grafana: docs/sources/tutorials/create-alerts-with-logs/index.md

    | ---------------- | ------------------------- |
    | Function         | Sum                       |
    | Input            | A                         |
    | Mode             | Strict                    |

    **Box C - threshold expression**
    |                  |                           |
    | ---------------- | --------------------------|
    | Input            | B                         |
    | Expression value | Is above 5                |
    | Alert condition  |This is the alert condition|

1. Expand **Options** and select **Range** as the type.

1. Click on **preview** to see a preview of the query result and alert evaluation. If you see **No Data** verify that you changed the query type to **Range**.

1. Expression B shows a table of labels and values returned. The message label captured the message string from the log line
 and the value shows the number of times that string occured during the evaluation interval.

    |   labels         |       values              |
    | ---------------- | --------------------------|
    |  message=level=info method=GET url=/ status=500 | 27 |
    |  message=level=info method=POST url=/ status=500 | 1 |

> 1. Configure your alert evaluation behavior

Reviewer: ```suggestion
1. Configure your alert evaluation behavior.
```

    * Choose a folder or use **+add new** to add a new folder for this alert.
    * Select an existing evaluation group from the drop-down or create a new one if this is your first alert.
    * Set the **for** value to **0s** so the alert will fire instantly.
    * Leave Configure no data and error handling No data handling on the default values.

1. Add an annotation that refers to labels and values from the query result in your alert notification.
  
    * Choose **+Add new** in the drop down and type the annotation name **AlertValues** into the blank box
    * In the blank `text` box paste ```{{ $labels.message }}  has returned an error status {{$values.B}} times.```

1. Click the **Save and exit** button at the top of the alert definition page

### Create a Loki managed alert

[Loki managed alerts](https://grafana.com/docs/loki/latest/rules/#alerting-and-recording-rules) are stored and evaluated by Loki. They use LogQL for their expressions. 

1. Choose  Mimir or Loki managed alert to create an alert using Loki

> 1. Select your Loki datasource from the drop-down

Reviewer: ```suggestion
1. Select your Loki data source from the drop-down
```

    * In Alerting under the Contact points tab:
        * Choose **Grafana** to use the built-in alertmanager
        * Click **+Add template**
        * Name the template `mynotification`
        * Add the snippet below to your alert template in the **Content** field. Notice that you will reference the annotation from your alert by name `(.Annotations.AlertValues)` to insert the annotation string into the alert notification:

        ```
        {{ define "myalert" }}
        [{{.Status}}] {{ .Labels.alertname }}
        {{ .Annotations.AlertValues }}
        {{ end }}
        {{ define "mymessage" }}
        {{ if gt (len .Alerts.Firing) 0 }}
            {{ len .Alerts.Firing }} firing:
            {{ range .Alerts.Firing }} {{ template "myalert" .}} {{ end }}
        {{ end }}
        {{ if gt (len .Alerts.Resolved) 0 }}
            {{ len .Alerts.Resolved }} resolved:
            {{ range .Alerts.Resolved }} {{ template "myalert" .}} {{ end }}
        {{ end }}
        {{ end }}
        ```
        * There are two sections to the notification template: 
            1. The `myalert` template creates a single alert notification based on a specific alert.

>             2. The `mymessage` template will find all of the grouped alerts that are firing and send them in a single notification.

Reviewer: ```suggestion
            1. The `mymessage` template will find all of the grouped alerts that are firing and send them in a single notification.
```

    ```
    chmod 755  ./web-server-logs-simulator.py
    ```

1. Run the script.

    * Use `tee` to direct the script output to the console and the specified file path. For example, if promtail is
    configured to monitor `/var/log` for `.log` files you can direct the script output to `/var/log/web_requests.log` file.

    * To avoid running the script with elevated permissions, create the log file manually and change the permissions for the output file only.

        ```
        sudo touch /var/log/web_requests.log
        chmod 755 /var/log/web_requests.log
        python3 ./web-server-logs-simulator.py | tee -a /var/log/web_requests.log
        ```

1. Verify that the logs are showing up in Grafana’s Explore view:

    * Navigate to explore in Grafana.
    * Select the Loki datasource from the drop-down.
    * Check the toggle for **builder | code** in the top right corner of the query box and switch the query mode to builder if it’s not already selected.
    * Select the filename label from the drop-down and choose your `web_requests.log` file from the value drop-down.
    * Click **Run Query**.
    * You should see logs and a graph of log volume.

> #### Troubleshooting the script

Reviewer: I think you want this to be either an H3 or H2, depending on whether it should be nested under `## Optional: Use promtail with a sample log-generating script` or on the same level, respectfully

If you don't see the sample logs in Explore:

* Does the output file exist, check /var/log/web_requests.log to see if it contains logs.
* If the file is empty, check that you followed the steps above to create the file and change the permissions.
* If the file exists, verify that promtail is running and check that it is configured correctly.
* In Grafana Explore, check that the time range is only for the last 5 minutes.

## Optional: Use Docker compose to create the tutorial environment 

These optional steps walk you through installing Grafana, Loki and Promtail with Docker compose. You'll also configure a log-generating script
that generates the sample logs used in this tutorial to create alerts.

> #### Pre-requisites

Reviewer: If you want this nested under the `## Optional...` heading above, make it an H3 (not H4)
